File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.py
import sys
import collections
sys.setrecursionlimit(10 ** 8)

N, M, R = map(int, sys.stdin.readline().split(" "))

graph = [[] for i in range(N + 1)]

visited = [False for i in range(N + 1)]

ans = [0 for i in range(N + 1)]

for i in range(M):
    a, b = map(int, sys.stdin.readline().split(" "))
    graph[a].append(b)
    graph[b].append(a)

for s in graph:
    s.sort()

seq = 0

def BFS(start):
    global seq
    seq += 1
    ans[start] = seq
    visited[start] = True
    q = collections.deque([start])
    while q:
        dot = q.popleft()
        for s in graph[dot]:
            if(visited[s] == False):
                visited[s] = True
                seq += 1
                ans[s] = seq
                q.append(s)

BFS(R)

for i in range(1, N + 1):
    print(ans[i])

File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.java
import java.util.*;
import java.io.*;

public class Main{

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static StringTokenizer st;
    private static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
    private static int[] answer;
    private static int seq = 0;

    private static void BFS(int start) {
        Deque<Integer> d = new LinkedList<>();
        seq += 1;
        answer[start] = seq;
        d.add(start);
        while(!d.isEmpty()) {
            int dot = d.poll();
            for(int i = 0 ; i < graph.get(dot).size() ; i++) {
                int newDot = graph.get(dot).get(i);
                if(answer[newDot] == 0) {
                    seq += 1;
                    answer[newDot] = seq;
                    d.add(newDot);
                }
            }
        }
    }


    public static void main(String[] args) throws IOException {

        st = new StringTokenizer(br.readLine(), " ");
        int N, M, R;
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());

        answer = new int[N + 1];

        for(int i = 0 ; i <= N ; i++) { 
            graph.add(new ArrayList<>());
        } 

        for(int i = 0 ; i < M ; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());

            graph.get(start).add(end);
            graph.get(end).add(start);
        }

        for(int i = 0 ; i < graph.size() ; i++) {
            Collections.sort(graph.get(i));
        }


        BFS(R);

        for(int i = 1 ; i <= N ; i++) {
            bw.write(Integer.toString(answer[i]));
            bw.write("\n");
        }

        bw.flush();

        br.close();
        bw.close();
    }
}